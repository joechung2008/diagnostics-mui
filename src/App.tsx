import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import {
  Button,
  Grid,
  Menu,
  MenuItem,
  Tab,
  Tabs,
  ThemeProvider,
  Toolbar,
} from "@mui/material";
import { useEffect, useMemo, useState } from "react";
import BuildInfo from "./BuildInfo";
import Extension from "./Extension";
import Extensions from "./Extensions";
import ServerInfo from "./ServerInfo";
import { useSystemTheme } from "./useSystemTheme";
import { isExtensionInfo } from "./utils";

import "@fontsource/roboto/300.css";
import "@fontsource/roboto/400.css";
import "@fontsource/roboto/500.css";
import "@fontsource/roboto/700.css";
import "./App.css";

const EnvironmentUrls = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type Environment = (typeof EnvironmentUrls)[keyof typeof EnvironmentUrls];

const App: React.FC = () => {
  const theme = useSystemTheme();

  const [diagnostics, setDiagnostics] = useState<Diagnostics>();
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [environment, setEnvironment] = useState<Environment>(
    EnvironmentUrls.Public
  );
  const [envMenuAnchorEl, setEnvMenuAnchorEl] = useState<null | HTMLElement>(
    null
  );
  const handleEnvMenuOpen = (event: React.MouseEvent<HTMLElement>) => {
    setEnvMenuAnchorEl(event.currentTarget);
  };
  const handleEnvMenuClose = () => {
    setEnvMenuAnchorEl(null);
  };
  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const environments = useMemo(
    () => [
      {
        key: "public",
        text: "Public Cloud",
        selected: environment === EnvironmentUrls.Public,
        onClick: () => {
          setEnvironment(EnvironmentUrls.Public);
          setExtension(undefined);
        },
      },
      {
        key: "fairfax",
        text: "Fairfax",
        selected: environment === EnvironmentUrls.Fairfax,
        onClick: () => {
          setEnvironment(EnvironmentUrls.Fairfax);
          setExtension(undefined);
        },
      },
      {
        key: "mooncake",
        text: "Mooncake",
        selected: environment === EnvironmentUrls.Mooncake,
        onClick: () => {
          setEnvironment(EnvironmentUrls.Mooncake);
          setExtension(undefined);
        },
      },
    ],
    [environment]
  );

  useEffect(() => {
    const getDiagnostics = async () => {
      const response = await fetch(environment);
      setDiagnostics(await response.json());
    };
    getDiagnostics();
  }, [environment]);

  if (!diagnostics) {
    return null;
  }

  const { buildInfo, extensions, serverInfo } = diagnostics;

  const handleLinkClick = (_?: React.MouseEvent, item?: KeyedNavLink) => {
    if (item) {
      const extension = extensions[item.key];
      if (isExtensionInfo(extension)) {
        setExtension(extension);
      }
    }
  };

  return (
    <ThemeProvider theme={theme}>
      <div className="flexbox">
        <Toolbar>
          <Button
            aria-controls={envMenuAnchorEl ? "env-menu" : undefined}
            aria-haspopup="true"
            endIcon={
              <ExpandMoreIcon
                sx={{
                  transform: envMenuAnchorEl
                    ? "rotate(180deg)"
                    : "rotate(0deg)",
                  transition: "transform 0.2s",
                }}
              />
            }
            onClick={handleEnvMenuOpen}
          >
            {environments.find((env) => env.selected)?.text ??
              "Select Environment"}
          </Button>
          <Menu
            id="env-menu"
            anchorEl={envMenuAnchorEl}
            open={Boolean(envMenuAnchorEl)}
            onClose={handleEnvMenuClose}
          >
            {environments.map((env) => (
              <MenuItem
                key={env.key}
                selected={env.selected}
                onClick={() => {
                  env.onClick();
                  handleEnvMenuClose();
                }}
              >
                {env.text}
              </MenuItem>
            ))}
          </Menu>
          {showPaasServerless && (
            <Button
              key="paasserverless"
              onClick={() => {
                const paasserverless =
                  diagnostics?.extensions["paasserverless"];
                if (isExtensionInfo(paasserverless)) {
                  setExtension(paasserverless);
                }
              }}
            >
              paasserverless
            </Button>
          )}
          <Button
            key="websites"
            onClick={() => {
              const websites = diagnostics?.extensions["websites"];
              if (isExtensionInfo(websites)) {
                setExtension(websites);
              }
            }}
          >
            websites
          </Button>
        </Toolbar>
        <Tabs
          value={selectedTab}
          onChange={(_, newValue) => setSelectedTab(newValue as string)}
          aria-label="App navigation"
        >
          <Tab label="Extensions" value="extensions" />
          <Tab label="Build Information" value="build" />
          <Tab label="Server Information" value="server" />
        </Tabs>
        {selectedTab === "extensions" && (
          <Grid className="tab-panel" container>
            <Grid className="stack" container gap="0.5rem">
              <Extensions
                extensions={extensions}
                onLinkClick={handleLinkClick}
              />
              {extension && <Extension {...extension} />}
            </Grid>
          </Grid>
        )}
        {selectedTab === "build" && (
          <div className="tab-panel">
            <BuildInfo {...buildInfo} />
          </div>
        )}
        {selectedTab === "server" && (
          <div className="tab-panel">
            <ServerInfo {...serverInfo} />
          </div>
        )}
      </div>
    </ThemeProvider>
  );
};

export default App;
