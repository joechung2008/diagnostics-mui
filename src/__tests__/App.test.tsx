import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { beforeEach, describe, expect, it, vi } from "vitest";
import App from "../App";

const diagnostics = {
  buildInfo: { buildVersion: "1.0.0" },
  extensions: {
    ext1: { extensionName: "ext1" },
  },
  serverInfo: {
    deploymentId: "deploy-1",
    extensionSync: { totalSyncAllCount: 1 },
    hostname: "localhost",
    nodeVersions: "v18.0.0",
    serverId: "server-1",
    uptime: 100,
  },
};

describe("App", () => {
  beforeEach(() => {
    vi.stubGlobal(
      "fetch",
      vi.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve(diagnostics),
        })
      )
    );
  });

  it("matches snapshot in loading state", () => {
    vi.stubGlobal(
      "fetch",
      vi.fn(() => new Promise(() => {})) // never resolves
    );
    const { asFragment } = render(<App />);
    // App returns null in loading state, so fragment should be empty
    expect(asFragment()).toMatchInlineSnapshot(`<DocumentFragment />`);
  });

  it("renders main UI after diagnostics fetch", async () => {
    render(<App />);
    await waitFor(() => {
      expect(screen.getByRole("tab", { name: /Extensions/i })).not.toBeNull();
      expect(
        screen.getByRole("tab", { name: /Build Information/i })
      ).not.toBeNull();
      expect(
        screen.getByRole("tab", { name: /Server Information/i })
      ).not.toBeNull();
    });
  });

  it("matches snapshot after diagnostics fetch", async () => {
    const { asFragment } = render(<App />);
    await waitFor(() => {
      expect(
        screen.getByRole("tab", { name: /Extensions/i })
      ).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot after switching to Build Information tab", async () => {
    const { asFragment } = render(<App />);
    await waitFor(() => {
      const tabs = screen.getAllByRole("tab");
      expect(tabs[1]).toBeInTheDocument(); // Build Information tab
    });
    const tabs = screen.getAllByRole("tab");
    fireEvent.click(tabs[1]); // Build Information tab
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot after switching to Server Information tab", async () => {
    const { asFragment } = render(<App />);
    await waitFor(() => {
      const tabs = screen.getAllByRole("tab");
      expect(tabs[2]).toBeInTheDocument(); // Server Information tab
    });
    const tabs = screen.getAllByRole("tab");
    fireEvent.click(tabs[2]); // Server Information tab
    expect(asFragment()).toMatchSnapshot();
  });
});
