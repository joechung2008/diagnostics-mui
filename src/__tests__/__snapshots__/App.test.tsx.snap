// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > matches snapshot after diagnostics fetch 1`] = `
<DocumentFragment>
  <div
    class="flexbox"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
    >
      <button
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Public Cloud
        <span
          class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-n5hn8a-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        websites
      </button>
    </div>
    <div
      class="MuiTabs-root css-jhczcq-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          aria-label="App navigation"
          class="MuiTabs-list MuiTabs-flexContainer css-hzcega-MuiTabs-list"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            Extensions
            <span
              class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Build Information
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Server Information
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row tab-panel css-d64g0g-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row stack css-r95vsg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
        >
          <ul
            class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
          >
            <div
              class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                >
                  ext1
                </span>
              </div>
            </div>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > matches snapshot after switching to Build Information tab 1`] = `
<DocumentFragment>
  <div
    class="flexbox"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
    >
      <button
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Public Cloud
        <span
          class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-n5hn8a-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        websites
      </button>
    </div>
    <div
      class="MuiTabs-root css-jhczcq-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          aria-label="App navigation"
          class="MuiTabs-list MuiTabs-flexContainer css-hzcega-MuiTabs-list"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            Extensions
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Build Information
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Server Information
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row tab-panel css-d64g0g-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row stack css-r95vsg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
        >
          <ul
            class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
          >
            <div
              class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                >
                  ext1
                </span>
              </div>
            </div>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > matches snapshot after switching to Server Information tab 1`] = `
<DocumentFragment>
  <div
    class="flexbox"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
    >
      <button
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Public Cloud
        <span
          class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-n5hn8a-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        websites
      </button>
    </div>
    <div
      class="MuiTabs-root css-jhczcq-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          aria-label="App navigation"
          class="MuiTabs-list MuiTabs-flexContainer css-hzcega-MuiTabs-list"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            Extensions
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Build Information
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Server Information
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row tab-panel css-d64g0g-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row stack css-r95vsg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
        >
          <ul
            class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
          >
            <div
              class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                >
                  ext1
                </span>
              </div>
            </div>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
